### PR TITLE
Add NFT Manager Positions

### PR DESCRIPTION
**Pull request by dpinones**
Closes: #123

### CODE REVIEW

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 1]
```diff
@@ -1,24 +1,635 @@
...
+            }
+            pool
+        }
+
+        fn yas_mint_callback(
+            ref self: ContractState, amount_0_owed: u256, amount_1_owed: u256, data: Array<felt252>
+        ) {
+            let msg_sender = get_caller_address();
+
+            // TODO: we need verify if data has a valid ContractAddress
```
Comment (rcatalan98) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:391]: Can we solve this TODO in this PR?

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 2]
```diff
@@ -1,24 +1,635 @@
...
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
+        ) {
+            let mut state = ERC721::unsafe_new_contract_state();
+            ERC721::ERC721Impl::safe_transfer_from(ref state, from, to, token_id, data);
+        }
+
+        // TODO: replace _approve
```
Comment (rcatalan98) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:462]: Why is this necessary?

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 3]
```diff
@@ -1,24 +1,635 @@
...
+    /// @return liquidity The amount of returned liquidity
+    fn get_liquidity_for_amount_1(
+        sqrt_ratio_AX96: FixedType, sqrt_ratio_BX96: FixedType, amount_1: u256
+    ) -> u128 {
+        let (sqrt_ratio_AX96, sqrt_ratio_BX96) = if sqrt_ratio_AX96 > sqrt_ratio_BX96 {
+            (sqrt_ratio_BX96, sqrt_ratio_AX96)
+        } else {
+            (sqrt_ratio_AX96, sqrt_ratio_BX96)
+        };
+        // TODO: .mag
```
Comment (rcatalan98) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:594]: What do you mean you want to return the `mag` of this result? There is no need for that in this case because you are already working with the `mag` right?

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 4]
```diff
@@ -1,24 +1,635 @@
...
+        pool_ids: LegacyMap<ContractAddress, u128>,
+        pool_id_to_pool_key: LegacyMap<u128, PoolKey>,
+        positions: LegacyMap<u256, Position>,
+        next_id: u256,
+        next_pool_id: u128
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, factory: ContractAddress) {
+        // TODO: ERC721Permit.sol
```
Comment (rcatalan98) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:212]: Aren't we going to avoid this implementation?

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+                );
+
+            let (min_tick, max_tick) = get_min_tick_and_max_tick();
+            yas_nft_position_manager
+                .mint(
+                    MintParams {
+                        token_0: token_0.contract_address,
+                        token_1: token_1.contract_address,
+                        fee: fee_amount(FeeAmount::MEDIUM),
+                        recipient: OTHER(), // TODO:
```
Comment (rcatalan98) [inline crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:212]: What's the missing TODO here?

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                        amount_0_desired: 15,
+                        amount_1_desired: 15,
+                        amount_0_min: 0,
+                        amount_1_min: 0,
+                        deadline: 10
+                    }
+                );
+
+            assert(yas_nft_position_manager.balance_of(OTHER()) == 1, 'wrong balance_of OTHER');
+            // assert(yas_nft_position_manager.token_of_owner_by_index(other.address, 0) == 1, ''); // TODO: ERC721Enumerable
```
Comment (rcatalan98) [inline crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:224]: What's the meaning of this assert?

**Review by rcatalan98 — CHANGES_REQUESTED** [review]

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                    }
+                );
+        }
+
+        #[test]
+        #[available_gas(200000000)]
+        fn test_creates_a_token() {
+            let (yas_nft_position_manager, token_0, token_1) = setup();
+
+            let encode_price_sqrt_1_1 = FP64x96Impl::new(79228162514264337593543950336, false);
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo]: I chose this number because it is the one used in the test case for [Uniswap Periphery](https://github.com/Uniswap/v3-periphery/blob/697c2474757ea89fec12a4e6db16a574fe259610/test/NonfungiblePositionManager.spec.ts#L218). We already have a function that returns this value and explains where it comes from

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 8]
```diff
@@ -1,24 +1,635 @@
...
-use yas_core::numbers::signed_integer::i32::i32;
+
+use yas_core::numbers::fixed_point::implementations::impl_64x96::FixedType;
+use yas_core::numbers::signed_integer::{i32::i32};
+
+// details about the uniswap position
+#[derive(Copy, Drop, Serde, starknet::Store)]
+struct Position {
+    // the nonce for permits
+    nonce: u128,
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: You're right, this variable is not necessary

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 2 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: You're right, the comment is not necessary. I've already removed it

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 4 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: You're right, this comment is not necessary since we won't be using that implementation anymore

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 5 above)
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo]: I removed this comment

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 8 above)
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 2 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 4 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: ✔️

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 5 above)
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo]: ✔️

File: crates/yas_core/src/contracts/yas_pool.cairo [hunk 9]
```diff
@@ -23,11 +36,13 @@ trait IYASPool<TContractState> {
     ) -> (u256, u256);
     fn token_0(self: @TContractState) -> ContractAddress;
     fn token_1(self: @TContractState) -> ContractAddress;
+    fn slot_0(self: @TContractState) -> (FixedType, i32, u8, bool);
```
Comment (uri-99) [inline crates/yas_core/src/contracts/yas_pool.cairo:39]: This implementation will collide with the one made in PR #135 :
`fn slot_0(self: @TContractState) -> Slot0;`

File: crates/yas_core/src/contracts/yas_pool.cairo [hunk 10]
```diff
@@ -199,6 +203,16 @@ mod YASPool {
             self.token_1.read()
         }
 
+        fn slot_0(self: @ContractState) -> (FixedType, i32, u8, bool) {
+            let slot_0 = self.slot_0.read();
+            (slot_0.sqrt_price_X96, slot_0.tick, slot_0.fee_protocol, self.unlocked.read())
+        }
```
Comment (uri-99) [inline crates/yas_core/src/contracts/yas_pool.cairo:209]: This implementation will collide with the one made in PR #135 :
`fn slot_0(self: @ContractState) -> Slot0 {
            self.slot_0.read()
        }`

**Review by uri-99 — COMMENTED** [review]
`fn slot_0()` implementation collision

File: crates/yas_core/src/tests/utils/constants.cairo [hunk 11]
```diff
@@ -75,6 +75,10 @@ mod PoolConstants {
         contract_address_const::<'WALLET'>()
     }
 
+    fn OTHER() -> ContractAddress {
+        contract_address_const::<'OTHER'>()
+    }
```
Comment (uri-99) [inline crates/yas_core/src/tests/utils/constants.cairo:80]: What does this constant represent?

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        let yas_factory = deploy_factory(OWNER(), POOL_CLASS_HASH()); // 0x1
+        let nft_position_manager = deploy_nft_position_manager(yas_factory.contract_address); // 0x2
+
+        // Deploy ERC20 tokens with factory address
+        let token_0 = deploy_erc20('YAS0', '$YAS0', 4000000000000000000, OWNER()); // 0x3
+        let token_1 = deploy_erc20('YAS1', '$YAS1', 4000000000000000000, OWNER()); // 0x4
+
+        set_contract_address(OWNER());
+        token_0.transfer(WALLET(), 4000000000000000000);
+        token_1.transfer(WALLET(), 4000000000000000000);
```
Comment (uri-99) [inline crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:42]: If the tests don't require a specific token limit, we could simply set it to `BoundedInt::max()` rather than the fixed value of 4x10¹⁸ . This approach was effective in resolving other issues in the past. Do you think this could apply here as well?

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    fn get_min_tick_and_max_tick() -> (i32, i32) {
+        let tick_spacing = IntegerTrait::<i32>::new(tick_spacing(FeeAmount::MEDIUM), false);
+        let min_tick = i32_div_no_round(MIN_TICK(), tick_spacing) * tick_spacing;
+        let max_tick = i32_div_no_round(MAX_TICK(), tick_spacing) * tick_spacing;
+        (min_tick, max_tick)
+    }
+
+    mod mint {
```
Comment (uri-99) [inline crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:100]: I think all previous test mods were PascalCase
```suggestion
    mod Mint {
```

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 14]
```diff
@@ -1,24 +1,630 @@
...
+            (liquidity, amount_0, amount_1, pool_dispatcher)
+        }
+    }
+
+    /// @notice Computes the amount of liquidity received for a given amount of token0 and price range
+    /// @dev Calculates amount0 * (sqrt(upper) * sqrt(lower)) / (sqrt(upper) - sqrt(lower))
+    /// @param sqrtRatioAX96 A sqrt price representing the first tick boundary
+    /// @param sqrtRatioBX96 A sqrt price representing the second tick boundary
+    /// @param amount0 The amount0 being sent in
+    /// @return liquidity The amount of returned liquidity
```
Comment (dubzn) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:559]:
```suggestion
    /// @dev Calculates amount0 * (sqrt(upper) * sqrt(lower)) / (sqrt(upper) - sqrt(lower))
    /// @param sqrt_ratio_AX96 A sqrt price representing the first tick boundary
    /// @param sqrt_ratio_BX96 A sqrt price representing the second tick boundary
    /// @param amount_0 The amount0 being sent in
    /// @return liquidity The amount of returned liquidity
```

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 15]
```diff
@@ -1,24 +1,630 @@
...
+        FullMath::mul_div(amount_0, intermediate, (sqrt_ratio_BX96 - sqrt_ratio_AX96).mag)
+            .try_into()
+            .unwrap()
+    }
+
+    /// @notice Computes the amount of liquidity received for a given amount of token1 and price range
+    /// @dev Calculates amount1 / (sqrt(upper) - sqrt(lower)).
+    /// @param sqrtRatioAX96 A sqrt price representing the first tick boundary
+    /// @param sqrtRatioBX96 A sqrt price representing the second tick boundary
+    /// @param amount1 The amount1 being sent in
```
Comment (dubzn) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:579]:
```suggestion
    /// @param sqrt_ratio_AX96 A sqrt price representing the first tick boundary
    /// @param sqrt_ratio_BX96 A sqrt price representing the second tick boundary
    /// @param amount_1 The amount1 being sent in
```

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 16]
```diff
@@ -1,24 +1,630 @@
...
+            .unwrap()
+    }
+
+    /// @notice Computes the maximum amount of liquidity received for a given amount of token0, token1, the current
+    /// pool prices and the prices at the tick boundaries
+    /// @param sqrtRatioX96 A sqrt price representing the current pool prices
+    /// @param sqrtRatioAX96 A sqrt price representing the first tick boundary
+    /// @param sqrtRatioBX96 A sqrt price representing the second tick boundary
+    /// @param amount0 The amount of token0 being sent in
+    /// @param amount1 The amount of token1 being sent in
```
Comment (dubzn) [inline crates/yas_periphery/src/yas_nft_position_manager.cairo:601]:
```suggestion
    /// @param sqrt_ratio_X96 A sqrt price representing the current pool prices
    /// @param sqrt_ratio_AX96 A sqrt price representing the first tick boundary
    /// @param sqrt_ratio_BX96 A sqrt price representing the second tick boundary
    /// @param amount_0 The amount of token0 being sent in
    /// @param amount_1 The amount of token1 being sent in
```

**Review by dubzn — CHANGES_REQUESTED** [review]

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 16 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:601]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 15 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:579]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 14 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:559]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 17]
```diff
@@ -1,24 +1,630 @@
...
+        self.next_pool_id.write(1);
+    }
 
     #[external(v0)]
     impl YASNFTPositionManagerImpl of IYASNFTPositionManager<ContractState> {
-        fn positions(self: @ContractState, token_id: u256) -> i32 {
-            IntegerTrait::<i32>::new(0, false)
+        fn factory(self: @ContractState) -> ContractAddress {
+            self.factory.read()
+        }
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:222]: This function belongs to the nftmanager interface. Rename the function to `get_factory`

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 18]
```diff
@@ -1,24 +1,630 @@
 use starknet::ContractAddress;
-use yas_core::numbers::signed_integer::i32::i32;
+
+use yas_core::numbers::fixed_point::implementations::impl_64x96::FixedType;
+use yas_core::numbers::signed_integer::{i32::i32};
+
+// details about the uniswap position
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:6]: ✔️

File: crates/yas_periphery/Scarb.toml [hunk 19]
```diff
@@ -4,6 +4,7 @@ version = "0.1.0"
 
 [dependencies]
 starknet.workspace = true
+openzeppelin.workspace = true
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/Scarb.toml:7]: It follows the same behavior as the starknet dependency. In the `scarb.toml` file located in the project's root, we have listed the dependencies. Now, if we want to use these dependencies in a workspace, the flag must be set to true.
Here are more details in the [scarb documentation](https://docs.swmansion.com/scarb/docs/reference/workspaces.html#dependencies).

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 13 above)
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:100]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 17 above)
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo [hunk 20]
```diff
@@ -1,24 +1,630 @@
...
+        to: ContractAddress,
+        token_id: u256,
+        data: Span<felt252>
+    );
+    fn approve(ref self: TContractState, to: ContractAddress, token_id: u256);
+    fn set_approval_for_all(ref self: TContractState, operator: ContractAddress, approved: bool);
+    fn get_approved(self: @TContractState, token_id: u256) -> ContractAddress;
+    fn is_approved_for_all(
+        self: @TContractState, owner: ContractAddress, operator: ContractAddress
+    ) -> bool;
```
[the remark this answers is not in the document]
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:121]: ✔️

File: crates/yas_core/src/contracts/yas_pool.cairo (hunk 10 above)
Comment (dpinones) [reply crates/yas_core/src/contracts/yas_pool.cairo]: ✔️

File: crates/yas_core/src/contracts/yas_pool.cairo (hunk 9 above)
Comment (dpinones) [reply crates/yas_core/src/contracts/yas_pool.cairo]: ✔️

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 12 above)
Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:42]: ✔️

File: crates/yas_periphery/src/yas_nft_position_manager.cairo (hunk 3 above)
Comment (dpinones) [reply crates/yas_periphery/src/yas_nft_position_manager.cairo:586]: ✔️

File: crates/yas_core/src/tests/utils/constants.cairo (hunk 11 above)
Comment (rcatalan98) [reply crates/yas_core/src/tests/utils/constants.cairo]: This is used to create contract address and the string between `''` is the key to create the test address

**Review by rcatalan98 — APPROVED** [review]
Great job!

File: crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo (hunk 12 above)
Comment (uri-99) [inline crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo:42]:
```suggestion
        let token_0 = deploy_erc20('YAS0', '$YAS0', BoundedInt::max(), OWNER()); // 0x3
        let token_1 = deploy_erc20('YAS1', '$YAS1', BoundedInt::max(), OWNER()); // 0x4

        set_contract_address(OWNER());
        token_0.transfer(WALLET(), BoundedInt::max());
        token_1.transfer(WALLET(), BoundedInt::max());
```

Comment (dpinones) [reply crates/yas_periphery/src/tests/test_yas_nft_position_manager.cairo]: ✔️

**Review by rcatalan98 — APPROVED** [review]

**Review by dubzn — APPROVED** [review]
